samplers/samplers.go

```diff
@@ -72,6 +73,131 @@ type JSONMetric struct {
 	Value []byte `json:"value"`
 }
 
+// CardinalityCountName is the name of the metric that we eventually pass on
+// to the sinks.
+const CardinalityCountName = "cardinality.count"
+
+// CardinalityCountType is the string that denotes the type of the metric. We
+// use this when we serialize to a JSON metric.
+const CardinalityCountType = "cardinality_count"
+
+// cardinalityPrecision is the precision of the HLL that stores cardinality for
+// each metric name.
+const cardinalityPrecision = 18
+
+// CardinalityCount is a sampler that records an approximate cardinality count
+// by metric name.
+type CardinalityCount struct {
+	MetricCardinality map[string]*hyperloglog.HyperLogLogPlus
+}
+
+func NewCardinalityCount() *CardinalityCount {
+	return &CardinalityCount{make(map[string]*hyperloglog.HyperLogLogPlus)}
+}
+
+// Sample adds a measurement to the cardinality counter. It makes the
+// assumption that no two metrics with different types share a name (ie, name
+// is unique across all types)
+func (cc *CardinalityCount) Sample(name string, joinedTags string) {
+	if _, present := cc.MetricCardinality[name]; !present {
+		hll, _ := hyperloglog.NewPlus(cardinalityPrecision)
+		cc.MetricCardinality[name] = hll
+	}
+	hasher := fnv.New64a()
```

Comment:
The `UDPMetric` has a `Digest` wherein [we've already computed the hash](https://github.com/stripe/veneur/blob/26f6de2ef55e7b1d40c428eee2136798b18ca5b9/samplers/parser.go#L207) of tags + type + name. Rather than rehashing you can just use that. Then your function can be `Sample(name string, digest uint32)`, your map keyed by the `digest` and you save a hash operation.